### PR TITLE
Enable private repositories

### PR DESCRIPTION
**Pull request by zetaab**
default behaviour is disabled. However, now it is possible to allow users to use private repositories as well.

fixes issue #36

### CODE REVIEW

**Comment by zetaab** [comment]
@jwilander could you review it. At least for us it works very well in github enterprise

**Comment by jwilander** [comment]
The only other cases I can think of is if you created a PR or issue in a repo then got removed and someone performed an action on it (comment, close, reopen, etc.). Looking at the possible notifications that could happen in most cases except being requested for a review or assigned (since you need to be in the repo for those). Might be safer to just check permissions in all cases for private repos

**Comment by jwilander** [comment]
Those changes look good! There's actually one other case I've thought of. Right now if you `/github subscribe somerepo` in a channel there's no check to see if you have permissions to `somerepo`. We'd likely need a permission check when using that command.

There's another concern I have with subscribing a channel to a private repository. If I have access to a repo and subscribe a channel to it and then I am removed from the private repo, my channel is still going get those events. I'm not entirely certain what the best way to handle this is. One way would be to track the creator of a subscription and before we post to the channel make sure that creator still has access to the repo. The downside there is if that user set up the subscription for a channel many users use, when that one user is removed the subscription stops working and the channel no longer gets the posts

Thoughts @zetaab or @lieut-data ?

**Comment by zetaab** [comment]
@jwilander if you do `/github subscribe somerepo` and you do not have access to repo - it will return "Unknown repository" like mentioned in issue #36. https://github.com/mattermost/mattermost-plugin-github/blob/master/server/subscriptions.go#L82-L86  So in my opinion there is no need to add permission check. 

Your another concern is true. However, I think that if user is so smart/stupid that he will add subscription from private repository to channel - he must know what he is doing.

**Comment by jwilander** [comment]
Ah right, I forgot about that for the first one.

For the second one, my worry is that the user could do it maliciously to maintain getting events about a repository that they were removed from

**Comment by jwilander** [comment]
@zetaab to address my second concern I think we need to:
1. Track who created the subscription
2. When a subscription event happens, check If the repo is private and then check if that the creator still has permission to the repo
3. If the user does not have permission, remove the subscription and post a message into the channel saying `@<username> does not have permission to <org/repo>. Subscription removed. Please run /github subscribe <org/repo> to recreate the subscription.`

**Comment by lieut-data** [comment]
Over to @levb and @crspeller given their ownership of all things plugins related now :)

**Comment by levb** [comment]
@zetaab I'd like to better familiarize myself github plugin code, will comment soon.

**Comment by jwilander** [comment]
I pushed a commit to handle the case I described. My only remaining worry is that for end users already using the GitHub plugin they will need to manually do `/github disconnect` then reconnect their account to get the correct OAuth scopes needed. Might be a good idea to provide a notification telling existing users they need to do that if they want to use the plugin with private repos

File: server/webhook.go [hunk 1]
```diff
@@ -158,6 +181,10 @@ func (p *Plugin) postPullRequestEvent(event *github.PullRequestEvent) {
 			continue
 		}
 
+		if isPrivate && !p.permissionToRepo(sub.CreatorID, repoName) {
```
Comment (crspeller) [inline server/webhook.go]: Is it possible to centeralize these checks? They all do similar things but are all spread out. Is it possible to just do the check in `handleWebhook`?

Comment (jwilander) [reply server/webhook.go]: Possibly with a significant amount of refactoring

**Comment by cpitclaudel** [comment]
This would be a great feature to have. Thanks everyone for working on it.

File: server/webhook.go (hunk 1 above)
Comment (jwilander) [reply server/webhook.go]: @crspeller pushed some refactoring that should clean it up a bit

**Comment by jwilander** [comment]
@crspeller pushed some refactoring that should clean it up a bit

**Comment by zetaab** [comment]
I cannot approve my own PR but looks great for me. Hopefully someone has tested it :)

**Comment by jwilander** [comment]
I tested it fairly thoroughly and we already have a volunteer to test a build of it before we make an official release :)

**Comment by levb** [comment]
I am the least qualified in the code base, should I hit the merge button?

**Comment by jwilander** [comment]
We're good to merge. Before I cut a new release I want to add the message to end users letting them know they need to reconnect their accounts to get private repos to work but I will do that in a new PR.

**Comment by cpitclaudel** [comment]
Great work!

**Pull request merged**